src/analysis/vector/qgsabstracttool.cpp

```diff
@@ -0,0 +1,110 @@
+ *  begin                : Jun 10, 2014                                    *
+ *  copyright            : (C) 2014 by Sandro Mani / Sourcepole AG         *
+ *  email                : smani@sourcepole.ch                             *
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QApplication>
+#include <QtConcurrentRun>
+#include <QtConcurrentMap>
+#include "qgsabstracttool.h"
+#include "qgsfield.h"
+#include "qgscoordinatereferencesystem.h"
+#include "qgsvectordataprovider.h"
+#include "qgsvectorfilewriter.h"
+#include "qgsvectorlayer.h"
+#include "qgsgeometry.h"
+#include "qgslogger.h"
+#include "qgswkbtypes.h"
+#include "qgsgeos.h"
+
+namespace Vectoranalysis
+{
+  void QgsAbstractTool::ProcessFeatureWrapper::operator()( const Job *job )
+  {
+    try
+    {
+      instance->processFeature( job );
+    }
+    catch ( const QgsException &e )
+    {
+      instance->mExceptions.append( e.what() );
+    }
+  }
+
+  QgsAbstractTool::QgsAbstractTool( QgsFeatureSink *output, QgsCoordinateTransformContext transformContext, QgsFeatureRequest::InvalidGeometryCheck invalidGeometryCheck ): mOutput( output ), mTransformContext( transformContext ), mInvalidGeometryCheck( invalidGeometryCheck )
+  {
+  }
+
+  QgsAbstractTool::~QgsAbstractTool()
+  {
+    qDeleteAll( mJobQueue );
+  }
+
+  QFuture<void> QgsAbstractTool::init()
+  {
+    return QtConcurrent::run( this, &QgsAbstractTool::prepare );
+  }
+
+  QFuture<void> QgsAbstractTool::execute()
+  {
+    return QtConcurrent::map( mJobQueue, ProcessFeatureWrapper( this ) );
+  }
+
+  void QgsAbstractTool::buildSpatialIndex( QgsSpatialIndex &index, QgsFeatureSource *layer ) const
+  {
+    QgsFeatureRequest request;
+    request.setFlags( QgsFeatureRequest::SubsetOfAttributes );
+    request.setSubsetOfAttributes( QgsAttributeList() );
+    QgsFeatureIterator it = layer->getFeatures( request );
+    index = QgsSpatialIndex( it );
+  }
+
+  void QgsAbstractTool::appendToJobQueue( QgsFeatureSource *layer, int taskFlag )
+  {
+    for ( const QgsFeatureId &id : layer->allFeatureIds() )
+    {
+      mJobQueue.append( new Job( id, taskFlag ) );
+    }
+  }
+
+  bool QgsAbstractTool::getFeatureAtId( QgsFeature &feature, QgsFeatureId id, QgsFeatureSource *layer, const QgsAttributeList &attIdx )
```

Comment:
These one-by-one requests are **extremely** slow for many feature sources, and must be avoided.

Reply:
Thank you for coming back to this PR. I agree the fetching one-by-one can be a problem, especially with slow remote DBs.
I have some ideas how to avoid the one-by-one fetching, going to  make some tests in the next days.